app/org/sagebionetworks/bridge/upload/IosSchemaValidationHandler.java

```diff
@@ -0,0 +1,559 @@
+    // be as flexible as possible. Which means our error handling strategy is to write a warning to the logs, and then
+    // attempt to recover. This will, however, cause cascading errors further down the validation chain.
+    @Override
+    public void handle(@Nonnull UploadValidationContext context)
+            throws UploadValidationException {
+        Map<String, JsonNode> jsonDataMap = context.getJsonDataMap();
+        Map<String, byte[]> unzippedDataMap = context.getUnzippedDataMap();
+        Upload upload = context.getUpload();
+        String uploadId = upload.getUploadId();
+        Study study = context.getStudy();
+        String studyId = study.getIdentifier();
+
+        // Add empty record builder and attachment map to the context. We'll fill these in as we need them.
+        HealthDataRecordBuilder recordBuilder = healthDataDao.getRecordBuilder();
+        context.setHealthDataRecordBuilder(recordBuilder);
+        Map<String, byte[]> attachmentMap = new HashMap<>();
+        context.setAttachmentsByFieldName(attachmentMap);
+
+        // health data records fields
+        recordBuilder.withHealthCode(upload.getHealthCode());
+        recordBuilder.withStudyId(studyId);
+        recordBuilder.withUploadDate(upload.getUploadDate());
+
+        // create an empty object node in our record builder, which we'll fill in as we go
+        ObjectNode dataMap = BridgeObjectMapper.get().createObjectNode();
+        recordBuilder.withData(dataMap);
+
+        JsonNode infoJson = jsonDataMap.get("info.json");
+        if (infoJson == null) {
+            // Recover by replacing this with an empty map
+            addMessageAndWarn(context, String.format("upload ID %s does not contain info.json file", uploadId));
+            infoJson = BridgeObjectMapper.get().createObjectNode();
+
+            // Add it back to the jsonDataMap, since all the logic assumes it contains info.json.
+            jsonDataMap.put("info.json", infoJson);
+        }
+
+        // Use info.json verbatim is the metadata.
+        recordBuilder.withMetadata(infoJson);
+
+        // extract other data from info.json
+        JsonNode taskRunNode = infoJson.get("taskRun");
+        String taskRunId = taskRunNode != null ? taskRunNode.textValue() : null;
+        JsonNode itemNode = infoJson.get("item");
+        String item = itemNode != null ? itemNode.textValue() : null;
+        
+        // Make sure all files specified by info.json are accounted for.
+        // Because ParseJsonHandler moves files from unzippedDataMap to jsonDataMap, there is no overlap between the
+        // two maps.
+        Set<String> fileNameSet = new HashSet<>();
+        fileNameSet.addAll(jsonDataMap.keySet());
+        fileNameSet.addAll(unzippedDataMap.keySet());
+
+        // fileList.size() should be exactly 1 less than fileNameSet.size(), because fileList.size() doesn't include
+        // info.json.
+        JsonNode fileList = infoJson.get("files");
```

Comment:
I know you're going to clean up later but it seems like a method that returns this JsonNode would encapsulate this stanza.